tests/test_syntax/blocks/test_html_blocks.py

```diff
@@ -0,0 +1,783 @@
+
+    def test_raw_p_no_end_tag(self):
+        self.assertMarkdownRenders(
+            '<p>*text*',
+            '<p>*text*'
+        )
+
+    def test_raw_multiple_p_no_end_tag(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <p>*text*'
+
+                <p>more *text*
+                """
+            ),
+            self.dedent(
+                """
+                <p>*text*'
+
+                <p>more *text*
+                """
+            )
+        )
+
+    def test_raw_p_no_end_tag_followed_by_blank_line(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <p>*raw text*'
+
+                Still part of *raw* text.
+                """
+            ),
+            self.dedent(
+                """
+                <p>*raw text*'
+
+                Still part of *raw* text.
+                """
+            )
+        )
+
+    def test_raw_nested_p_no_end_tag(self):
+        self.assertMarkdownRenders(
+            '<div><p>*text*</div>',
+            '<div><p>*text*</div>'
+        )
+
+    def test_raw_open_bracket_only(self):
+        self.assertMarkdownRenders(
+            '<',
+            '<p>&lt;</p>'
+        )
+
+    def test_raw_open_bracket_followed_by_space(self):
+        self.assertMarkdownRenders(
+            '< foo',
+            '<p>&lt; foo</p>'
+        )
+
+    def test_raw_missing_close_bracket(self):
+        self.assertMarkdownRenders(
+            '<foo',
+            '<p>&lt;foo</p>'
+        )
+
+    def test_raw_attributes(self):
+        self.assertMarkdownRenders(
+            '<p id="foo", class="bar baz", style="margin: 15px; line-height: 1.5; text-align: center;">text</p>',
+            '<p id="foo", class="bar baz", style="margin: 15px; line-height: 1.5; text-align: center;">text</p>'
+        )
+
+    def test_raw_attributes_nested(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <div id="foo, class="bar", style="background: #ffe7e8; border: 2px solid #e66465;">
+                    <p id="baz", style="margin: 15px; line-height: 1.5; text-align: center;">text</p>
+                </div>
+                """
+            ),
+            self.dedent(
+                """
+                <div id="foo, class="bar", style="background: #ffe7e8; border: 2px solid #e66465;">
+                    <p id="baz", style="margin: 15px; line-height: 1.5; text-align: center;">text</p>
+                </div>
+                """
+            )
+        )
+
+    def test_raw_comment_one_line(self):
+        self.assertMarkdownRenders(
+            '<!-- *foo* -->',
+            '<!-- *foo* -->'
+        )
+
+    # TODO: Decide behavior here. Python-Markdown current outputs:
+    #
+    #   <!-- *foo* -->
+    #   <p><em>bar</em></p>
+    #
+    # But the reference implementation outputs:
+    #
+    #   <p><!-- *foo* --><em>bar</em></p>
```

Comment:
Good point.

Reply:
So, with CDATA you suggest treating the tag as inline, but here you suggest treating the tag as block-level. In both cases, the reason for the question is the test after the tag on the same line. It seems to me all tags should exhibit the same behavior in that situation.

So my question is: if an otherwise block-level tag is not alone on a line(s), do we still treat it as block-level or inline? I have the same questiosn on multiple senarios in these tests, and would prefer the same behavior on all (unless there is a good argument to do otherwise).

Reply:
I was not considering this as the same case as CDATA, I did not have in mind that these two cases are linked together.

> if an otherwise block-level tag is not alone on a line(s), do we still treat it as block-level or inline?

The implementations divide into two large groups (and one small group) here: [results for `<div>foo</div> bar`](https://johnmacfarlane.net/babelmark2/?normalize=1&text=%3Cdiv%3Efoo%3C%2Fdiv%3E+bar).

If there is no clear leader, then maybe we should stick to what the reference implementation (Markdown.pl 1.0.2b8) does? I.e. treat the element as block-level but then process the remaining part of line separately.